src/libraries/Common/src/System/Net/Security/SslKeyLogger.cs

```diff
@@ -0,0 +1,132 @@
+            {
+                s_fileStream = File.Open(s_keyLogFile, FileMode.Append, FileAccess.Write, FileShare.ReadWrite);
+            }
+        }
+        catch (Exception ex)
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(null, $"Failed to open SSL key log file '{s_keyLogFile}': {ex}");
+            }
+        }
+    }
+#pragma warning restore CA1810
+
+    public static bool IsEnabled => s_fileStream != null;
+
+    public static void WriteLineRaw(ReadOnlySpan<byte> data)
+    {
+        Debug.Assert(s_fileStream != null);
+        if (s_fileStream == null)
+        {
+            return;
+        }
+
+        if (data.Length > 0)
+        {
+            lock (s_fileStream)
+            {
+                s_fileStream.Write(data);
+                s_fileStream.WriteByte((byte)'\n');
+                s_fileStream.Flush();
+            }
+        }
+    }
+
+    public static void WriteSecrets(
+        ReadOnlySpan<byte> clientRandom,
+        ReadOnlySpan<byte> clientHandshakeTrafficSecret,
+        ReadOnlySpan<byte> serverHandshakeTrafficSecret,
+        ReadOnlySpan<byte> clientTrafficSecret0,
+        ReadOnlySpan<byte> serverTrafficSecret0,
+        ReadOnlySpan<byte> clientEarlyTrafficSecret)
+    {
+        Debug.Assert(s_fileStream != null);
+        Debug.Assert(!clientRandom.IsEmpty);
+
+        if (s_fileStream == null ||
+            clientRandom.IsEmpty ||
+
+            // return early if there is nothing to log
+            (clientHandshakeTrafficSecret.IsEmpty &&
+            serverHandshakeTrafficSecret.IsEmpty &&
+            clientTrafficSecret0.IsEmpty &&
+            serverTrafficSecret0.IsEmpty &&
+            clientEarlyTrafficSecret.IsEmpty))
+        {
+            return;
+        }
+
+        Span<byte> clientRandomUtf8 = clientRandom.Length <= 1024 ? stackalloc byte[clientRandom.Length * 2] : new byte[clientRandom.Length * 2];
+        HexEncode(clientRandom, clientRandomUtf8);
+
+        lock (s_fileStream)
+        {
+            WriteSecretCore("CLIENT_HANDSHAKE_TRAFFIC_SECRET"u8, clientRandomUtf8, clientHandshakeTrafficSecret);
+            WriteSecretCore("SERVER_HANDSHAKE_TRAFFIC_SECRET"u8, clientRandomUtf8, serverHandshakeTrafficSecret);
+            WriteSecretCore("CLIENT_TRAFFIC_SECRET_0"u8, clientRandomUtf8, clientTrafficSecret0);
+            WriteSecretCore("SERVER_TRAFFIC_SECRET_0"u8, clientRandomUtf8, serverTrafficSecret0);
+            WriteSecretCore("CLIENT_EARLY_TRAFFIC_SECRET"u8, clientRandomUtf8, clientEarlyTrafficSecret);
+
+            s_fileStream.Flush();
+        }
+    }
+
+    private static void WriteSecretCore(ReadOnlySpan<byte> labelUtf8, ReadOnlySpan<byte> clientRandomUtf8, ReadOnlySpan<byte> secret)
+    {
+        if (secret.Length == 0)
+        {
+            return;
+        }
+
+        int totalLength = labelUtf8.Length + 1 + clientRandomUtf8.Length + 1 + 2 * secret.Length + 1;
```

Comment:
Why does this need to be super optimized? 
The original seemed much easier to see what will actually end up in the output:
```
$"CLIENT_EARLY_TRAFFIC_SECRET {clientRandom} {Convert.ToHexString(new ReadOnlySpan<byte>(_tlsSecrets->ClientEarlyTrafficSecret, _tlsSecrets->SecretLength))}\n"
```

Reply:
It probably doesn't, but since we are exposing it in the published product code and it is not that difficult to remove the allocation I felt it didn't hurt.

I will add a comment to make it easier to understand.